### ISSUE TITLE
Discriminated union intersected with discriminant does not join to the original type when discriminated

### ISSUE THREAD

**Issue by jack-williams** [microsoft/TypeScript#36333]
**TypeScript Version:**  master

**Search Terms:**

Discriminated union intersection narrowing

**Code**

```ts
type A = { key: "one", value: number } | { key: "two", value: string };

type A_one = A & { key: "one" };

function foo(a1: A_one): number {
    if (a1.key) {

    }
    return a1.value;  // no error
}

function bar(a1: A_one): number {
    return a1.value;  // error
}
```

**Expected behavior:**
Consistent behavior between `foo` and `bar`, though I'm not sure how easy that is.
- Making them both error feels like a regression.
- Making them both ok would need to introduce some new logic for arbitrary property access.

**Actual behavior:**
The control flow through the joint point is doing some type reduction not present in the plain version.

**Playground Link:** 

[link](http://www.typescriptlang.org/play/?ts=3.8.0-dev.20200119&ssl=1&ssc=1&pln=14&pc=2#code/C4TwDgpgBAglC8UDeUDWEQC4oCID2AdhDgDRQBuAhgDYCuE2BtAtgEYQBOUAvlAD7I0GbDmAB3PKQo162AM7AOASwIBzHgG4AUFtCRYAfULREcAGSD0WXMZyadAM1oEAxsCWEoDvHgAUlAEZsGCMiAEpGFnYuJC0oeKglByh-AIA6KzDkHQSeOISOCGBaDgIoQLSqOggNeIB6OqgCPChODjwOLW5HZzcPMtZKDlTg0IgIpqjObNzC4tLy9Kr6WqgG1o52zu4gA)

**Related Issues:** 

Found when looking at #36323.

This was (probably) introduced by the fix at #35513.

**Comment by AnyhowStep** [comment]
I think I've experience this problem a few times in the past, is why I don't use intersections anymore, for this kind of narrowing. I use `Extract<>`.

The workaround I'd use is,
```ts
type A_one = Extract<A, { key: "one" }>;
```

**Comment by RyanCavanaugh** [comment]
Fixed by #36696

**Comment by RumataEstor** [comment]
Could you please confirm it's the same issue that I see in the following and the fix works for it too?

```ts
type Good = { data: string };
type Bad = { error: string };

type NoError<T> = T & { error?: undefined };

type Either = NoError<Good> | Bad;

function main(either: Either) {
    if (either.error) {
        console.log("error", either.error);
    } else {
        console.log("data", either.data); // property data doesn't exist on type Bad
    }
}
```
